Refactor pkg/subctl/cmd/gather -> internal/gather

Jaanki opened the pull request:
This PR does the following:
1. Adds `config` to `cluster.Info`
2. Uses execute.OnMultiCluster functionality
3. Uses constants defined in internal/constants.go

Related to https://github.com/submariner-io/submariner-operator/issues/1329

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
3. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
5. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
6. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
7. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
8. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

Jaanki committed:
Add config to cluster.Info API

and use new cluster.Info API and constants

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

tpantelis reviewed (commented):
So it looks like you copied code to _internal/gather_  but it's not being used.

Jaanki commented:
> So it looks like you copied code to _internal/gather_ but it's not being used.

Yes. e2e is failing for some reason so didn't push more commits.

Jaanki commented on pkg/cluster/info.go:
```diff
@@ -28,17 +28,20 @@ import (
+func NewInfo(clusterName string, clientProducer client.Producer, config *rest.Config) (*Info, error) {
 	info := &Info{
 		Name:           clusterName,
+		RestConfig:     config,
```
`RestConfig` is needed by `gatherCNIResources` -> `execCmdInBash` -> `pods.ExecWithOprions` leading to https://github.com/submariner-io/submariner-operator/blob/devel/internal/pods/exec.go#L90

Jaanki committed:
Use internal/gather

.. and delete cobra stuff from internal/gather/gather.go

Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>

tpantelis reviewed (changes requested):
To maintain git history, we should first move the files to _internal/gather_.

dfarrell07 commented:
> To maintain git history, we should first move the files to _internal/gather_.

We have a pointer for this guidance at https://submariner.io/development/code-review/#rename-and-edit-in-separate-commits FYI. I think it was created because of very similar feedback on similar refactoring a few months ago.

Jaanki commented:
> To maintain git history, we should first move the files to _internal/gather_.

`git mv` moves the files from `pkg/subctl/cmd/gather/.` to `internal/gather`. I want the files to be **copied** and **not moved.**

tpantelis commented:
> git mv moves the files from pkg/subctl/cmd/gather/. to internal/gather. I want the files to be copied and not moved.

Why? Copying loses the git history. 

```
git log -- internal/gather/cni.go

commit d11b28efd3454e88ec8a69826270366bb4f24133
Author: Janki Chhatbar <jchhatba@redhat.com>
Date:   Fri Apr 22 11:45:28 2022 +0530

    Copy pkg/subctl/cmd/gather -> internal/gather
    
    Signed-off-by: Janki Chhatbar <jchhatba@redhat.com>
```

The deceiving thing tho is that the diff in the GtHub UI shows that it was renamed w/o changes.

skitt commented on internal/gather/gather.go:
```diff
@@ -0,0 +1,270 @@
+type Options struct {
+	Directory            string
+	IncludeSensitiveData bool
+	ModuleFlags          map[string]bool
+	TypeFlags            map[string]bool
```
These should be string slices, not maps; that would make it easier to use them as Cobra flags directly (and we don’t care about the `bool`).

tpantelis replied:
Commit https://github.com/submariner-io/submariner-operator/pull/2007/commits/0e5e8a46eeb0ac3155f9f0a85d70626850fd1c56.

skitt commented on internal/gather/gather.go:
```diff
@@ -0,0 +1,270 @@
+	Resources = "resources"
+)
+
+var ModuleFlags = map[string]bool{
+	component.Connectivity:     false,
+	component.ServiceDiscovery: false,
+	component.Broker:           false,
+	component.Operator:         false,
+}
```
This shouldn’t be necessary, the same information is effectively present in `gatherFuncs`; perhaps a `GetSupportedModules() []string` function could be used instead, or an `IsModuleSupported(module string) bool` function.

tpantelis replied:
Made it a set with https://github.com/submariner-io/submariner-operator/pull/2007/commits/0e5e8a46eeb0ac3155f9f0a85d70626850fd1c56.

skitt commented on internal/gather/gather.go:
```diff
@@ -0,0 +1,270 @@
+	component.Operator:         false,
+}
+
+var TypeFlags = map[string]bool{
+	"logs":      false,
+	"resources": false,
+}
```
Likewise, this should be hidden behind a function, and maintained as a `[]string`.

tpantelis replied:
Made it a set with https://github.com/submariner-io/submariner-operator/pull/2007/commits/0e5e8a46eeb0ac3155f9f0a85d70626850fd1c56

tpantelis committed:
Move pkg/subctl/cmd/gather -> internal/gather

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis committed:
Use string slices for gather Types and Modules

... instead of maps and use them directly as Cobra flags directly
via StringSliceVar.

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

skitt commented on cmd/subctl/diagnose.go:
```diff
@@ -210,7 +210,7 @@ func clusterInfoFromKubeConfig(kubeConfig string) *cluster.Info {
 	exit.OnErrorWithMessage(err, fmt.Sprintf("Error creating client producer for kubeconfig %q", kubeConfig))
 
-	clusterInfo, err := cluster.NewInfo("", clientProducer)
+	clusterInfo, err := cluster.NewInfo("", clientProducer, nil)
```
This is unfortunate, but let’s not block the PR — see #2042.

tpantelis replied:
Agree.
